File: Code/Model/Domain Objects/Option.swift
//
//  Option.swift
//  TH-PopQuiz
//
//  Created by Ryne Cheow on 8/3/15.
//  Copyright (c) 2015 TradeHero. All rights reserved.
//

import Argo
import Runes

class Option: JSONDecodable, DebugPrintable, Equatable, DictionaryRepresentation {
    let content : String
    let accessoryImageUrl : String?

    var imageContent: UIImage!

    lazy var isGraphical: Bool = {
        return self.accessoryImageUrl != nil
    }()

    init(content:String, accessoryImageUrl:String?){
        self.content = content
        self.accessoryImageUrl = accessoryImageUrl
    }
    
    class func create(content: String)(accessoryImageUrl: String?)-> Option {
        return Option(content: content, accessoryImageUrl: accessoryImageUrl)
    }
    
    class func decode(j: JSONValue) -> Option? {
        return Option.create
            <^> j <| "content"
            <*> j <| "accessoryImageUrl"
    }

    /**
     Initialise option with string content and image content

     :param: stringContent The string content of the option
     */
    
    func fetchImage(completionHandler: () -> ()) {
        if let imgName = self.accessoryImageUrl {
            NetworkClient.fetchImageFromURLString(imgName, progressHandler: nil, completionHandler: {
                image, error in
                if error != nil {
                    debugPrintln(error)
                    return
                }
                
                if image != nil {
                    self.imageContent = image
                }
                
                completionHandler()
            })

        } else {
            completionHandler()
        }
    }

    var debugDescription: String {
        return "[ Option content: \(content) ]"
    }
    
    var dictionaryRepresentation: [String:AnyObject] {
        return ["content": content, "accessoryImageUrl": accessoryImageUrl ?? NSNull()];
    }
}

func ==(lhs: Option, rhs: Option) -> Bool {
    return lhs.content == rhs.content
}

File: HTML5/src/model/Option.js
(function () {
    /**
     * convert json of option to object
     * @param content
     * @constructor
     */
    PopQuiz.Option = function Option(content) {
        /**
         *
         * @type {string}
         */
        this.stringContent = "";
        /**
         *
         * @type {string}
         */
        this.imageContentURLString = "";
        /**
         *
         * @returns {boolean}
         */
        this.isGraphical = function () {
            return this.imageContentURLString !== "";
        };

        var split = content.split("|");
        this.stringContent = split[0];
        this.imageContentURLString = "";
        if (split.length === 2) {
            this.imageContentURLString = split[1];
        }
    };
}() );